bots/flakes-refresh

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+
+# This file is part of Cockpit.
+#
+# Copyright (C) 2017 Red Hat, Inc.
+#
+# Cockpit is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import sys
+import time
+import os
+import urllib
+import socket
+import json
+
+sys.dont_write_bytecode = True
+
+import task
+
+NUMBER_OPEN_ISSUES = 7            # How many issues do we want to have open at a given time?
+
+# How far back does our data go?  If a flake gets fixed but is still
+# flaky after this many days, the bots open another issue.
+
+WINDOW_DAYS = 90
+
+# This parses the output JSONL format discussed here, where various
+# values are grouped:
+#
+# https://github.com/cockpit-project/cockpituous/blob/master/learn/README.md
+
+# Here we're looking for a field in a record that only has one value
+def value(record, field):
+    values = record.get(field, [])
+    if len(values) == 1:
+        return values[0][0] or ""
+    return None
+
+# Here we're looking for the count of a specific field/value in the record
+def count(record, field, only):
+    values = record.get(field, [])
+    for value, count in values:
+        if value != only:
+            continue
+        return count
+    return 0
+
+def run(api, verbose=False, dry=False):
+    open_issues = api.issues(labels=[ "flake" ])
+    create_count = NUMBER_OPEN_ISSUES - len(open_issues)
+
+    if create_count <= 0:
+        return 0
+
+    if verbose:
+        sys.stderr.write("Going to create %s new flake issue(s)\n" % create_count)
+
+    host = os.environ.get("COCKPIT_LEARN_SERVICE_HOST")
+    if not host:
+        raise RuntimeError("no learn host environment variable: $COCKPIT_LEARN_SERVICE_HOST")
+    port = os.environ.get("COCKPIT_LEARN_SERVICE_PORT", "8080")
+    url = "http://{0}:{1}/active/statistics.jsonl".format(host, port)
+
+    statistics = [ ]
+
+    # Retrieve the URL
+    try:
+        req = urllib.request.Request(url)
+        with urllib.request.urlopen(req) as f:
```

Comment:
Again harmless, but normally one directly specifies the URL in urlopen(), not a Request object.

Reply:
Changed.